prog.go

```diff
@@ -242,14 +242,24 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
 	insns := make(asm.Instructions, len(spec.Instructions))
 	copy(insns, spec.Instructions)
 
-	handle, fib, lib, err := btf.MarshalExtInfos(insns)
-	if err != nil && !errors.Is(err, btf.ErrNotSupported) {
-		return nil, fmt.Errorf("load ext_infos: %w", err)
+	var b btf.Builder
+	if err := applyRelocations(insns, opts.KernelTypes, spec.ByteOrder, &b); err != nil {
+		return nil, fmt.Errorf("apply CO-RE relocations: %w", err)
 	}
-	if handle != nil {
-		defer handle.Close()
 
-		attr.ProgBtfFd = uint32(handle.FD())
+	if !b.Empty() && haveProgramExtInfos() != nil {
+		// Return ErrNotSupported instead of E2BIG if CO-RE relies on BTF
+		// being present.
+		return nil, haveProgramExtInfos()
+	}
+
+	if haveProgramExtInfos() == nil {
```

Comment:
Maybe it's time to move this to a function? `newProgram` is getting crowded.

Reply:
I'll take another look!

Reply:
Yes, it needs a refactor. But I don't have the time at the moment.